icgcget/dispatcher.py

```diff
@@ -1,192 +1,256 @@
-    logger = logging.getLogger('__name__')
-    if 'cghub' in object_ids and object_ids['cghub']:
-        cli.check_access(cghub_access, 'cghub')
-        gt_client = GnosDownloadClient()
-        return_code = gt_client.download(object_ids['cghub'], cghub_access, cghub_path, output,
-                                         cghub_transport_parallel)
-        cli.check_code('Cghub', return_code)
-
-    if 'aws-virginia' in object_ids and object_ids['aws-virginia']:
-        cli.check_access(icgc_access, 'icgc')
-        icgc_client = StorageClient()
-        return_code = icgc_client.download(object_ids['aws-virginia'], icgc_access, icgc_path, output,
-                                           icgc_transport_parallel, file_from=icgc_transport_file_from, repo='aws')
-        cli.check_code('Icgc', return_code)
-
-    if 'ega' in object_ids and object_ids['ega']:
-        cli.check_access(ega_access, 'ega')
-        if ega_transport_parallel != '1':
-            logger.warning("Parallel streams on the ega client may cause reliability issues and failed " +
-                           "downloads.  This option is not recommended.")
-        ega_client = EgaDownloadClient()
-        return_code = ega_client.download(object_ids['ega'], ega_access, ega_path, output, ega_transport_parallel,
-                                          ega_udt)
-        cli.check_code('Ega', return_code)
-
-    if 'collaboratory' in object_ids and object_ids['collaboratory']:
-        cli.check_access(icgc_access, 'icgc')
-        icgc_client = StorageClient()
-        return_code = icgc_client.download(object_ids['collaboratory'], icgc_access, icgc_path, output,
-                                           icgc_transport_parallel, file_from=icgc_transport_file_from, repo='collab')
-        cli.check_code('Icgc', return_code)
-
-    if 'gdc' in object_ids and object_ids['gdc']:
-        cli.check_access(gdc_access, 'gdc')
-        gdc_client = GdcDownloadClient()
-        return_code = gdc_client.download(object_ids['gdc'], gdc_access, gdc_path, output, gdc_transport_parallel,
-                                          gdc_udt)
-        cli.check_code('Gdc', return_code)
-
-
-def status_tables(repos, fileids, manifest, api_url, no_files):
-    logger = logging.getLogger('__log__')
-    repo_list = []
-    gdc_ids = []
-    cghub_ids = []
-    repo_sizes = {}
-    repo_counts = {}
-    repo_donors = {}
-    donors = []
-    type_donors = {}
-    type_sizes = {}
-    type_counts = {}
-    total_size = 0
-
-    file_table = [["", "Size", "Unit", "File Format", "Data Type", "Repo"]]
-    summary_table = [["", "Size", "Unit", "File Count", "Donor_Count"]]
-    if manifest:
-        if len(fileids) > 1:
-            logger.warning("For download from manifest files, multiple manifest id arguments is not supported")
-            raise click.BadArgumentUsage("Multiple manifest files specified.")
+class Dispatcher:
```

Comment:
Sure. Have a look around this project:

https://github.com/aws/aws-cli/tree/develop/awscli

It is very similar in structure, although more complex, to what we are trying to do. Notice the `customization` package. Pretty these are the "commands". We shouldn't take this design too literally, but I think it is a good reference point for the project in terms of decomposition.

I would need to think a bit more about our project and study the code before making a suggestion. For now, just keep this in the back of your mind.